airgun/entities/domain.py

```diff
@@ -0,0 +1,192 @@
+    DomainEditView
+)
+
+
+class DomainEntity(BaseEntity):
+    _DEFAULT_ORG = 'Default Organization'
+    _DEFAULT_LOC = 'Default Location'
+    _CREATE_SUCCESSFUL = 'Successfully created {}.'
+    _UPDATE_SUCCESSFUL = 'Successfully updated {}.'
+    _DELETE_SUCCESSFUL = 'Successfully deleted {}.'
+
+    @staticmethod
+    def _submit_domain_form(view, **kwargs):
+        """Create or update domain.
+
+        Form is the same for both cases, we use 1 method here to fill the form.
+        """
+        dns_domain = kwargs.get('dns_domain')
+        full_name = kwargs.get('full_name')
+        dns_capsule = kwargs.get('dns_capsule')
+        parameters = kwargs.get('parameters')
+        locations = kwargs.get('locations')
+        organizations = kwargs.get('organizations')
+
+        # Fill 'Domain' tab
+        if dns_domain:
+            view.domain.dns_domain.fill(dns_domain)
+        if full_name:
+            view.domain.full_name.fill(full_name)
+        if dns_capsule:
+            view.domain.dns_capsule.fill(dns_capsule)
+
+        # Fill 'Parameters' tab
+        if parameters:
+            params = []
+            for param_name, param_value in parameters.items():
+                params.append({'name': param_name, 'value': param_value})
+            view.parameters.params.fill(params)  # fill overwrites old params
+
+        # Fill locations tab
+        if locations:
+            view.locations.multiselect.set_assigned(locations)
+
+        # Fill organizations tab
+        if organizations:
+            view.organizations.multiselect.set_assigned(organizations)
+
+        view.submit_button.click()
+
+    def create(self, dns_domain, **kwargs):
+        """
+        Create a new domain.
+
+        :param dns_domain: str, full DNS domain name
+        :param full_name: (optional) str, full name describing the domain
+        :param dns_capsule: (optional) str of DNS capsule to use
+        :param parameters: (optional) param dict or list of param dicts to set
+            Format: [{'name': str, 'value': str}]
+        :param locations: list of location strings to be assigned.
+            All other locations will be unassigned.
+            Default: ['Default Location']
+        :param organizations: list of org strings to be assigned.
+            All other orgs will be unassigned.
+            Default: ['Default Organization']
+        """
+        view = self.navigate_to(self, 'New')
+
+        kwargs['dns_domain'] = dns_domain
+
+        if not kwargs.get('locations'):
+            kwargs['locations'] = [self._DEFAULT_LOC]
+        if not kwargs.get('organizations'):
+            kwargs['organizations'] = [self._DEFAULT_ORG]
+
+        self._submit_domain_form(view, **kwargs)
+
+        wait_for(
+            lambda: view.flash.messages, fail_condition=[], timeout=10,
+            delay=2, message='wait for flash success'
+        )
+        view.flash.assert_success_message(self._CREATE_SUCCESSFUL.format(dns_domain))
```

Comment:
so you always expect that you procedure should be finished successfully which is also wrong as we need to perform a lot of negative testing, so that validation is also needs to be performed on test layer or at least we should be able to say what we want to expect in the end

Reply:
Personally i'd left flash notification type check here (assert_no_errors) - that one will still allow us to do negative testing and won't require any text constants in airgun.
Negative test will just ensure exception is risen after creation:
```python
with raises(AssertionError):
    session.domain.create('bad_inputs')
```
Take a look at https://github.com/SatelliteQE/airgun/blob/master/airgun/entities/activationkey.py#L45 , i'm checking whether notification appeared and whether it's a good one.

Regarding messages testing - that's a different story, you've suggested one of options but there probably are others. Let's discuss that in separate issue when we actually hit the test which requires such message check. AFAIK they're not numerous. In any case - i agree with @oshtaier - we want to isolate layers and not to introduce exact expected messages (which may change quite often) in airgun, they should lie in robottelo.